backend/src/main.rs

```diff
@@ -111,24 +113,127 @@ async fn main() -> anyhow::Result<()> {
                     .map_err(anyhow::Error::from),
                 None => Ok(()),
             }
         };
 
-        futures::try_join!(shutdown_signal, server_f, workers_f, monitor_f, metrics_f)?;
+        futures::try_join!(shutdown_signal, server_f, metrics_f, workers_f, monitor_f)?;
+    }
+    Ok(())
+}
+
+pub fn monitor_db(
+    db: &Pool<Postgres>,
+    client: windmill_api_client::Client,
+    timeout: i32,
+    rx: tokio::sync::broadcast::Receiver<()>,
+) {
+    let db1 = db.clone();
+    let db2 = db.clone();
+
+    let rx2 = rx.resubscribe();
+
+    tokio::spawn(async move {
+        windmill_worker::handle_zombie_jobs_periodically(&db1, &client, timeout, rx).await
+    });
+    tokio::spawn(async move { windmill_api::delete_expired_items_perdiodically(&db2, rx2).await });
+}
+
+pub async fn run_workers(
+    db: Pool<Postgres>,
+    addr: SocketAddr,
+    timeout: i32,
+    num_workers: i32,
+    sleep_queue: u64,
+    worker_config: WorkerConfig,
+    rx: tokio::sync::broadcast::Receiver<()>,
+) -> anyhow::Result<()> {
+    let instance_name = rd_string(5);
+    let monitor = tokio_metrics::TaskMonitor::new();
+
+    let ip = windmill_common::external_ip::get_ip()
+        .await
+        .unwrap_or_else(|e| {
+            tracing::warn!(error = e.to_string(), "failed to get external IP");
+            "unretrievable IP".to_string()
+        });
+
+    let mut handles = Vec::with_capacity(num_workers as usize);
+
+    for i in 1..(num_workers + 1) {
+        let db1 = db.clone();
+        let instance_name = instance_name.clone();
+        let worker_name = format!("dt-worker-{}-{}", &instance_name, rd_string(5));
+        let ip = ip.clone();
+        let rx = rx.resubscribe();
+        let worker_config = worker_config.clone();
+        let client = windmill_api_client::create_client(
+            &worker_config.base_url,
+            get_token(&db, &worker_name).await,
+        );
+        handles.push(tokio::spawn(monitor.instrument(async move {
+            tracing::info!(addr = %addr.to_string(), worker = %worker_name, "starting worker");
+            windmill_worker::run_worker(
+                &db1,
+                &client,
+                timeout,
+                &instance_name,
+                worker_name,
+                i as u64,
+                num_workers as u64,
+                &ip,
+                sleep_queue,
+                worker_config,
+                rx,
+            )
+            .await
+        })));
     }
 
+    futures::future::try_join_all(handles).await?;
     Ok(())
 }
+
+async fn get_token(db: &sqlx::Pool<sqlx::Postgres>, worker_name: &str) -> String {
+    // TODO: This is bad. Proper authentication should happen.
+    use rand::prelude::*;
+    let token: String = rand::thread_rng()
+        .sample_iter(&rand::distributions::Alphanumeric)
+        .take(30)
+        .map(char::from)
+        .collect();
+    let expiration = chrono::Utc::now() + chrono::Duration::days(2);
```

Comment:
What would happen currently after 2 days ? If things just break, let's make things perpetual and TODO implement some kind of refresh token logic

Reply:
Since the client is non super_admin, it cannot access workspace specific data and jobs should be part of that, i need to undestand more why it works at all

Reply:
I need to check again, but I think this token is only used for very few calls anyways. The most important one is the create_token, creating the per-job token, that does also have access to specific workspace.

Reply:
Creating per-job token based on a token should not be able to escalate token privilege by having a user that is not part of a workspace able to create a user-specific token part of a workspace. So the token created should either not have enough privilege or we've implemented the endpoint very very wrong.